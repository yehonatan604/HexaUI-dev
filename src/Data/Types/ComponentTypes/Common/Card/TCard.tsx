import { TComponent, ComponentOptions } from "../../TComponent";
import { Tsize } from "../../../TSize";

export type TCard = TComponent<"div"> & {
  options?: CardOptions;
};

export type CardOptions = ComponentOptions & {
  padding?: string;
  rounded?: Tsize | string;
  gap?: number;
};
