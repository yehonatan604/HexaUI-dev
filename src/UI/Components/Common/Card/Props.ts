import { TComponent } from "../../../../Data/Types/TComponent";

export type CardProps = TComponent<"div"> & {
    shadow?: boolean;
};