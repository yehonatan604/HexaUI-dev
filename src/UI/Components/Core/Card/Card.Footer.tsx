import { FlexDir } from "../../../../Data/Constants/FlexDirection";
import { FlexTypes } from "../../../../Data/Constants/FlexTypes";
import { TComponent } from "../../../../Data/Types/TComponent";
import Flex from "../../Layout/Flex/Flex";

const CardFooter = (props: TComponent<"div">) => {
  return (
    <Flex
      options={{
        direction: FlexDir.Col,
      }}
      className="w-full"
      {...props}
    >
      <hr className="my-2 border-stone-400 w-5/6 m-auto" />
      <Flex
        options={{
          align: FlexTypes.Center,
          justify: FlexTypes.SpaceEvenly,
        }}
        className="w-full p-3"
      >
        {props.children}
      </Flex>
    </Flex>
  );
};

export default CardFooter;
