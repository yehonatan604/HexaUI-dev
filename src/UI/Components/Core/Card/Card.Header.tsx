import useTheme from "../../../../Core/Context/ThemeContext/hooks/useTheme";
import { FlexDir } from "../../../../Data/Constants/FlexDirection";
import { TComponent } from "../../../../Data/Types/TComponent";
import Flex from "../../Layout/Flex/Flex";

const CardHeader = (props: TComponent<"div">) => {
  const { children, options, ...componentProps } = props;

  // Hooks
  const { colors } = useTheme();

  //Options
  const textColor = options?.textVariant || colors.standradLight;

  return (
    <Flex
      options={{
        direction: FlexDir.Col,
      }}
      {...componentProps}
    >
      <h1 className={`text-2xl text-${textColor} font-bold py-2`}>{children}</h1>
    </Flex>
  );
};

export default CardHeader;
