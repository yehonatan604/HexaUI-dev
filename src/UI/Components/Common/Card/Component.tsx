import useTheme from "../../../../Core/Hooks/useTheme";
import { FlexDir } from "../../../../Data/Constants/FlexDirection";
import { FlexTypes } from "../../../../Data/Constants/FlexTypes";
import { TCard } from "../../../../Data/Types/ComponentTypes/Common/Card/TCard";
import Flex from "../../Layout/Flex/Flex";
import CardFooter from "./Card.Footer/Component";
import CardHeader from "./Card.Header/Component";

const Card = (props: TCard) => {
  const { children, options, className, ...componentProps } = props;
  const { colors } = useTheme();

  // Options
  const bgColor = options?.bgVariant || colors.standradLight;
  const textColor = options?.textVariant || colors.standradDark;
  const padding = options?.padding || 2;
  const rounded = options?.rounded || "md";
  const gap = options?.gap || 4;
  const borderColor = options.border?.variant || colors.standrad;
  const borderRadius = options.border?.radius || "md";

  // Border Options
  const border = options.border?.show
    ? `border-2 border-${borderColor} rounded-${borderRadius}`
    : "";

  // Shadow Options
  const shadow =
    options?.shadow &&
    `shadow-${options?.shadow?.size || "md"} shadow-${
      options?.shadow?.color || colors.white
    }`;

  // JSX
  return (
    <Flex
      options={{
        direction: FlexDir.Col,
        justify: FlexTypes.Start,
        align: FlexTypes.Start,
      }}
      className={`gap-${gap} text-${textColor} p-${padding} rounded-${rounded} bg-${bgColor} ${shadow} ${border} ${className}`}
      {...componentProps}
    >
      {children}
    </Flex>
  );
};

Card.Header = CardHeader;
Card.Footer = CardFooter;
export default Card;
