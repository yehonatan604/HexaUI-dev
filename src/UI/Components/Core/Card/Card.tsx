import useTheme from "../../../../Core/Context/ThemeContext/hooks/useTheme";
import { FlexDir } from "../../../../Data/Constants/FlexDirection";
import { FlexTypes } from "../../../../Data/Constants/FlexTypes";
import Flex from "../../Layout/Flex/Flex";
import CardFooter from "./Card.Footer";
import CardHeader from "./Card.Header";
import { TCard } from "./types/TCard";

const Card = (props: TCard) => {
  const { children, options, className, ...componentProps } = props;
  const { colors } = useTheme();

  // Options
  const bgColor = options?.bgVariant || colors.standradLight;
  const textColor = options?.textVariant || colors.standradDark;
  const padding = options?.padding || 2;
  const rounded = options?.rounded || "md";
  const gap = options?.gap || 4;
  const borderColor = options?.border?.variant || colors.standrad;
  const borderRadius = options?.border?.radius || "md";

  // Border Options
  const border = options?.border?.show
    ? `border-2 border-${borderColor} rounded-${borderRadius}`
    : "";

  // Shadow Options
  const shadow =
    options?.shadow &&
    `shadow-${options?.shadow?.size || "md"} shadow-${
      options?.shadow?.color || colors.white
    }`;

  // JSX
  return (
    <Flex
      options={{
        direction: FlexDir.Col,
        justify: FlexTypes.Start,
        align: FlexTypes.Start,
      }}
      className={`gap-${gap} text-${textColor} p-${padding} rounded-${rounded} bg-${bgColor} ${shadow} ${border} ${className}`}
      {...componentProps}
    >
      {children}
    </Flex>
  );
};

Card.Header = CardHeader;
Card.Footer = CardFooter;
export default Card;
