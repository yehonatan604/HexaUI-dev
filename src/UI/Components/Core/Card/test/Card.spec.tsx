import { render, screen } from "@testing-library/react";
import { test, expect, describe } from "vitest";
import Card from "../Card";

describe("Card Component", () => {
  test("matches snapshot", () => {
    render(<Card data-testid="Card" />);
    expect(screen.getByTestId("Card")).toMatchSnapshot();
  });

  test("renders Card Header correctly", () => {
    render(<Card.Header>Card Header Title</Card.Header>);
    expect(screen.getByText("Card Header Title")).toBeInTheDocument();
  });

  test("renders Card Footer with button correctly", () => {
    render(
      <Card.Footer>
        <button>Submit</button>
      </Card.Footer>
    );
    expect(screen.getByText("Submit")).toBeInTheDocument();
  });

  test("renders Card with Header, Body, and Footer", () => {
    render(
      <Card>
        <Card.Header>Card Header</Card.Header>
        <div>Card Body Content</div>
        <Card.Footer>
          <button>Submit</button>
        </Card.Footer>
      </Card>
    );

    expect(screen.getByText("Card Header")).toBeInTheDocument();
    expect(screen.getByText("Card Body Content")).toBeInTheDocument();
    expect(screen.getByText("Submit")).toBeInTheDocument();
  });

  test("renders Card with custom options", () => {
    render(
      <Card
        data-testid="Card"
        options={{
          bgVariant: "red-500",
          textVariant: "white",
          rounded: "lg",
          gap: 2,
          border: {
            show: true,
            variant: "blue-500",
            radius: "xl",
          },
          shadow: {
            size: "lg",
            color: "gray-500",
          },
        }}
      >
        <Card.Header>Card Header</Card.Header>
        <div>Card Body Content</div>
        <Card.Footer>
          <button>Submit</button>
        </Card.Footer>
      </Card>
    );

    const card = screen.getByTestId("Card");
    expect(card).toHaveClass(
      "bg-red-500 text-white rounded-lg gap-2 shadow-lg shadow-gray-500 border-2 border-blue-500 rounded-xl"
    );
  });
});
