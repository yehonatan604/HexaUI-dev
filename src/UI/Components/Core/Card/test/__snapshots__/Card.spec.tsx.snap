// Vitest Snapshot v1

exports[`matches snapshot 1`] = `
<div
  class="flex flex-col bg- text- justify-start items-start flex-nowrap gap-4 text-standard-d p-2 rounded-md bg-standard-l undefined  undefined"
  data-testid="Card"
/>
`;
